## EJERCICIO FINAL

El fichero datos_pago_microcreditos.csv contiene 30000 transacciones bancarias relativas al pago o impago de microcreditos. El dataset consta de 62 dimensiones propietarias de las cuales el banco no ha proporcionado informacion por confidencialidad, a parte de su valor. La matriz de entrada se encuentra en x_train, e y_train contiene la etiqueta relativa a esa transaccion, un 1 indica que si se pago el microcredito y un 0 que no se pago.

Considera que el tamaño del dataset es lo suficientemente grande como para, dependiendo de los recursos de la maquina, poder tardar varios minutos en entrenar modelos complejos. 

Ejecuta el codigo inicial para estandarizar los dartos y contesta las preguntas.

In [90]:
# importamos las librerias necesarias
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [ ]:
# leemos el dataset
data_file='Project\data\datos_pago_microcreditos.csv'
pdl=pd.read_csv(data_file)

In [92]:
pdl.head(10)


payment var1 var2  var3   var4   var5   var6   var7   var8 var9  ...  \
0  Success   qw   hk  3.11  16.06  -4.60  22.34  13.53   1.53   nv  ...   
1   Denied   qw   rv  3.35  11.18 -18.55   6.68  12.78   6.62   nv  ...   
2   Denied   qw   zg  4.15  29.19  18.91  16.40   3.67   5.72   ch  ...   
3  Success   wv   js  6.23  15.70   2.81   4.46   5.13   8.66   ja  ...   
4  Success   ma   xn  1.28  20.71  14.98  11.19  17.66   1.13   nv  ...   
5  Success   kq   py -2.45  22.45  15.18  -2.12  -8.24  10.34   ch  ...   
6  Success   wv   py  1.05  23.02  17.59   6.65  -2.06  12.20   ch  ...   
7  Success   qw   rv  5.41  17.92 -14.59   5.00   1.34  -8.54   ch  ...   
8   Denied   wv   rv  7.29  26.83  33.92  13.35  20.57   4.46   ch  ...   
9   Denied   qw   rv  3.13  34.21  22.55   0.80   6.91  11.18   ch  ...   

   var21  var22  var23  var24  var25  var26  var27 var28  var29  var30  
0   8.94 -12.76     ub  12.06   2.46   4.73  -1.72  0.91     ev   8.00  
1  31.02  34.76     cz   1.44   9.44  13.56  -2.24  0.24     ev  -2.90  
2  23.26   9.50     ri   7.77   8.70  -1.75   5.96  1.91     ev  22.67  
3  29.25  -1.53     ri   8.94  19.33  23.73   5.54  0.85     ev  36.31  
4   2.19  10.24     ub   8.92   5.48  -0.28   4.01  1.21     ev  11.33  
5  -0.26   8.11     ri   2.32   6.89  -3.51   6.65 -1.12     ev   9.93  
6  14.39  -4.73     tf   8.57   8.73   2.87   1.22  1.26     ev  21.75  
7   4.09  22.62     ub   2.82 -11.05  11.94   9.08  0.68     ev   7.53  
8  12.25  21.17     tf  15.00   7.42  12.93   0.87  1.60     dg  -4.66  
9  14.06  14.27     tf   6.59   6.22   3.33  -2.54 -1.43     ev  13.69  

[10 rows x 31 columns]

In [93]:
print(pdl['payment'].unique())

['Success' 'Denied']


In [94]:
print(pdl.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 31 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   payment  30000 non-null  object 
 1   var1     30000 non-null  object 
 2   var2     30000 non-null  object 
 3   var3     30000 non-null  float64
 4   var4     30000 non-null  float64
 5   var5     30000 non-null  float64
 6   var6     30000 non-null  float64
 7   var7     30000 non-null  float64
 8   var8     30000 non-null  float64
 9   var9     30000 non-null  object 
 10  var10    30000 non-null  object 
 11  var11    30000 non-null  object 
 12  var12    30000 non-null  float64
 13  var13    30000 non-null  object 
 14  var14    30000 non-null  float64
 15  var15    30000 non-null  float64
 16  var16    30000 non-null  float64
 17  var17    30000 non-null  object 
 18  var18    30000 non-null  float64
 19  var19    30000 non-null  object 
 20  var20    30000 non-null  float64
 21  var21    300

In [95]:
print(f'El numero de filas es: {pdl.shape[0]}')

El numero de filas es: 30000


In [96]:
print(pdl.describe())

               var3          var4          var5          var6          var7  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean       3.018848     13.941943     19.981561      8.009453      9.110358   
std        3.004458     14.028019     19.866781      8.013395      8.976009   
min       -8.190000    -43.660000    -55.100000    -23.630000    -27.580000   
25%        0.990000      4.450000      6.520000      2.550000      3.050000   
50%        3.040000     13.920000     19.980000      8.010000      9.170000   
75%        5.040000     23.520000     33.452500     13.430000     15.152500   
max       15.160000     76.780000    107.080000     44.330000     48.390000   

               var8         var12         var14         var15         var16  \
count  30000.000000  30000.000000  30000.000000  30000.000000  30000.000000   
mean       5.993785      7.020662      2.007760     17.118747     10.919426   
std        6.035204      7.004504      2.001967    

In [97]:
# remlazamos los valores de la columna payment por 1 y 0 de acuerdo a las reglas del establecidadas
pdl["payment"]=np.where(pdl["payment"]=="Success",1,0)

k=pdl.columns # listamos las columnas del dataset

# modificamos las varialbes categoricas a dummies
for col in k:
    if pdl[col].dtype=='object':
        temp=pd.get_dummies(pdl[col],drop_first=True,prefix=col)
        pdl=pd.concat([pdl,temp],axis=1)
        pdl.drop([col],axis=1,inplace=True)
# eliminamos las filas que tienen valores nulos
pdl.dropna(axis=0,inplace=True)

In [98]:
pdl.info() # verificamos que no hay valores nulos y la info de las columnas 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 63 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   payment   30000 non-null  int64  
 1   var3      30000 non-null  float64
 2   var4      30000 non-null  float64
 3   var5      30000 non-null  float64
 4   var6      30000 non-null  float64
 5   var7      30000 non-null  float64
 6   var8      30000 non-null  float64
 7   var12     30000 non-null  float64
 8   var14     30000 non-null  float64
 9   var15     30000 non-null  float64
 10  var16     30000 non-null  float64
 11  var18     30000 non-null  float64
 12  var20     30000 non-null  float64
 13  var21     30000 non-null  float64
 14  var22     30000 non-null  float64
 15  var24     30000 non-null  float64
 16  var25     30000 non-null  float64
 17  var26     30000 non-null  float64
 18  var27     30000 non-null  float64
 19  var28     30000 non-null  float64
 20  var30     30000 non-null  fl

In [99]:
pdl.head(10) # mostramos las primeras 10 filas del dataset

payment  var3   var4   var5   var6   var7   var8  var12  var14  var15  ...  \
0        1  3.11  16.06  -4.60  22.34  13.53   1.53   4.46   4.93  26.48  ...   
1        0  3.35  11.18 -18.55   6.68  12.78   6.62   4.04  -0.76  16.21  ...   
2        0  4.15  29.19  18.91  16.40   3.67   5.72  -4.41   1.21  29.06  ...   
3        1  6.23  15.70   2.81   4.46   5.13   8.66   2.14   3.56  23.61  ...   
4        1  1.28  20.71  14.98  11.19  17.66   1.13  14.93   2.20 -19.16  ...   
5        1 -2.45  22.45  15.18  -2.12  -8.24  10.34   8.64   0.37  35.83  ...   
6        1  1.05  23.02  17.59   6.65  -2.06  12.20  14.91  -0.84  -8.38  ...   
7        1  5.41  17.92 -14.59   5.00   1.34  -8.54  13.97   5.85  28.99  ...   
8        0  7.29  26.83  33.92  13.35  20.57   4.46  15.19   5.23  -7.39  ...   
9        0  3.13  34.21  22.55   0.80   6.91  11.18  22.57   0.62  -0.84  ...   

   var23_da  var23_fe  var23_po  var23_qu  var23_ri  var23_sy  var23_tf  \
0     False     False     False     False     False     False     False   
1     False     False     False     False     False     False     False   
2     False     False     False     False      True     False     False   
3     False     False     False     False      True     False     False   
4     False     False     False     False     False     False     False   
5     False     False     False     False      True     False     False   
6     False     False     False     False     False     False      True   
7     False     False     False     False     False     False     False   
8     False     False     False     False     False     False      True   
9     False     False     False     False     False     False      True   

   var23_ub  var23_yv  var29_ev  
0      True     False      True  
1     False     False      True  
2     False     False      True  
3     False     False      True  
4      True     False      True  
5     False     False      True  
6     False     False      True  
7      True     False      True  
8     False     False     False  
9     False     False      True  

[10 rows x 63 columns]

In [100]:
pdl_train, pdl_test = train_test_split(pdl, test_size = 0.2,random_state=2) # dividimos el dataset en train y test, 80% y 20% respectivamente

#x_train=pdl_train.drop(["payment"],1)
x_train=pdl_train.drop(columns="payment") # variable independiente
y_train=pdl_train["payment"] # variable objetivo

#x_test=pdl_test.drop(["payment"],1)
x_test=pdl_test.drop(columns="payment") # variable independiente
y_test=pdl_test["payment"] # variable objetivo

x_train.reset_index(drop=True,inplace=True) # reiniciamos el index
y_train.reset_index(drop=True,inplace=True) # reiniciamos el index

In [101]:
y_train

0        0
1        1
2        0
3        0
4        1
        ..
23995    0
23996    1
23997    0
23998    0
23999    1
Name: payment, Length: 24000, dtype: int64

In [102]:
x_train

var3   var4   var5   var6   var7   var8  var12  var14  var15  var16  \
0      1.02  22.08   8.69  -9.24  21.06   2.21   1.33   2.00   7.62  -5.77   
1      2.09  11.25  23.46  12.11  29.57   4.77   8.86   2.38  29.48  28.14   
2      6.77  25.95   3.95  17.74  19.78   9.02   3.76   0.72  34.00  17.19   
3      2.75  12.47  -4.54  -1.42 -10.32   9.11   8.37   7.91   4.86   3.67   
4      4.47   7.90  31.60   2.53 -12.66   3.64   1.87   5.07  19.04  28.70   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
23995 -1.28   9.09  11.89   5.58  17.29  15.65   8.28   0.77  23.06   9.79   
23996  1.60   9.52  14.64   6.11  20.14  17.93   0.50   1.19   1.14   8.22   
23997  0.54  22.37  42.39  12.41   3.54   5.09   0.22   1.45  24.50   4.45   
23998  3.37  11.72  17.49  -2.63   0.83  12.47  17.58   0.32   3.22  11.61   
23999  5.21  28.45  27.84  14.82   2.13   0.82  -0.73   2.27  -4.48 -18.34   

       ...  var23_da  var23_fe  var23_po  var23_qu  var23_ri  var23_sy  \
0      ...     False      True     False     False     False     False   
1      ...     False     False     False     False     False     False   
2      ...     False     False     False     False     False     False   
3      ...     False     False     False     False     False     False   
4      ...     False     False     False     False     False     False   
...    ...       ...       ...       ...       ...       ...       ...   
23995  ...     False     False     False     False     False     False   
23996  ...     False     False     False     False     False     False   
23997  ...     False      True     False     False     False     False   
23998  ...     False     False     False     False     False      True   
23999  ...     False     False     False     False     False     False   

       var23_tf  var23_ub  var23_yv  var29_ev  
0         False     False     False     False  
1          True     False     False     False  
2         False      True     False      True  
3         False      True     False      True  
4          True     False     False      True  
...         ...       ...       ...       ...  
23995     False      True     False      True  
23996      True     False     False      True  
23997     False     False     False      True  
23998     False     False     False      True  
23999      True     False     False      True  

[24000 rows x 62 columns]

# Pregunta 1:
Construye un  clasificador usando K-NN con 3 vecinos y otro usando un arbol de decision. ¿cual produce mejor resultado? ¿que metrica has usado? segun el arbol de decision, ¿que metricas son las mas relevantes?


### 1. Clasificaddor K-NN

In [ ]:
# llamamos la libreria
from sklearn.neighbors import KNeighborsClassifier

# organizamos el modelo de KNN con 3 vecinos
knn = KNeighborsClassifier(n_neighbors=3).fit(x_train, y_train) # entrenamos el modelo

# evaluamos el modelo
print(f'El accuracy del modelo train es: {knn.score(x_train, y_train)}') 
print(f'El accuracy del modelo test es: {knn.score(x_test, y_test)}') 

El accuracy del modelo train es: 0.8305833333333333
El accuracy del modelo test es: 0.665


### 2. Clasificador Arbol de decision

In [ ]:
# importamos la libreria para el modelo de arbol de decision
from sklearn.tree import DecisionTreeClassifier

tree = DecisionTreeClassifier().fit(x_train, y_train)

# evaluamos el modelo
print(f'El accuracy del modelo train es: {tree.score(x_train, y_train)}')
print(f'El accuracy del modelo test es: {tree.score(x_test, y_test)}')


El accuracy del modelo train es: 1.0
El accuracy del modelo test es: 0.7821666666666667


Usando el accuracy (exactitud) se puede definir que:
* A simple vista el clasificador K-NN produce mejor resultado, obteniendo un accuracy en la particion de train de 0,83 y de 0,65 en test
* Mientras que el clasificaor de arbol de desicion entra en sobreajuste con 1.0 de accuracy sobre el train y 0,78 sobre test
    *  Esto se podria ajustar disminuyendo la complejidad del arbol, definiendo la profundidad del mismo

# Pregunta 2:
Repite el ejercicio anterior usando normalizacion de los datos y compara los resultados.

In [105]:
from sklearn.preprocessing import MinMaxScaler # llamamos la libreria para normalziar
scaler = MinMaxScaler() 

# aplicamos la normalizacion unicamente a las x
x_train_scaled = scaler.fit_transform(x_train) # normalizamos los datos de entrenamiento
x_test_scaled = scaler.transform(x_test) # normalizamos los datos de test

### 1. Clasificador K-NN con normalizacion de los datos

In [ ]:
knn = KNeighborsClassifier(n_neighbors=3).fit(x_train_scaled, y_train)

# evaluamos el modelo
print(f'El accuracy del modelo train es: {knn.score(x_train_scaled, y_train)}')
print(f'El accuracy del modelo test es: {knn.score(x_test_scaled, y_test)}')

El accuracy del modelo train es: 0.79925
El accuracy del modelo test es: 0.6168333333333333


Resultados K-NN sin normalizar datos
* El accuracy del modelo train es: 0.8305833333333333
* El accuracy del modelo test es: 0.665

### 2. Clasificador Arbol de desicion con normalizacion

In [ ]:
Tree = DecisionTreeClassifier().fit(x_train_scaled, y_train) 

# evaluamos el modelo
print(f'El accuracy del modelo train es: {Tree.score(x_train_scaled, y_train)}')
print(f'El accuracy del modelo test es: {Tree.score(x_test_scaled, y_test)}')

El accuracy del modelo train es: 1.0
El accuracy del modelo test es: 0.7821666666666667


Para esta ocasion aplicando normalizacion de los datos se tiene que: 
* El modelo con arbol de desicion sigue en overfitting (sobreajuste) debido a que su accuracy de train sigue igual en 1.0 
* Por otro lado el clasificador K-NN con los datos normalizados se ajusta mas a la generalizacion del modelo
    * Disminuye mi valor de train pero tambien disminuye mi valor de test

# Pregunta 3:
Usando GridSearchCV, identifica para un clasificador K-NN el numero de vecinos entre 1 y 30 que optimiza el resultado usando como scoring la precision y con 10 folds para la validacion cruzada. 

In [ ]:
from sklearn.model_selection import GridSearchCV
import numpy as np

# definimos el rango de vecinos de 1 a 30
param_grid = {'n_neighbors': np.arange(1, 31, 1)}

# eonfiguramos GridSearchCV con 10 folds y el clasificador K-NN
grid_search = GridSearchCV(
    KNeighborsClassifier(),
    param_grid=param_grid,
    scoring='accuracy',  # usamos el scoring de precision
    cv=10  # 10 dobles
)

# entrenamos el modelo
grid_search.fit(x_train, y_train)


print(f'el mejor promedio de accuracy con validacion cruzada es: {grid_search.best_score_:.3f}')
print(f'El mejor parametro de numero de vecinos es: {grid_search.best_params_}')

# evaluamos el modelo en el conjunto de test
print(f'El accuracy del modelo test es: {grid_search.score(x_test, y_test):.3f}')

el mejor promedio de accuracy con validacion cruzada es: 0.722
El mejor parametro de numero de vecinos es: {'n_neighbors': np.int64(27)}
El accuracy del modelo test es: 0.723


# Pregunta 4:
Obten la matriz de confusion del clasificador optimo anterior. ¿como lees cada uno de los valores?¿que valor de precission y recall tiene el clasificador optimizado con precission?

In [109]:
# importamos la libreria
from sklearn.metrics import ConfusionMatrixDisplay

In [110]:
from sklearn.metrics import confusion_matrix, classification_report, precision_score, recall_score

#  predicciones del clasificador anterior
y_pred = grid_search.best_estimator_.predict(x_test)

# matriz de confusión
conf_matrix = confusion_matrix(y_test, y_pred)
print("Matriz de Confusión:")
print(conf_matrix)

# precisión y recall
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)

print(f'\nPrecisión del clasificador optimizado anteriormente: {precision:.3f}')
print(f'Recall del clasificador optimizado: {recall:.3f}')

# Reporte de clasificación para más detalles
print('\nReporte de Clasificación:')
print(classification_report(y_test, y_pred))

Matriz de Confusión:
[[3424  338]
 [1323  915]]

Precisión del clasificador optimizado anteriormente: 0.730
Recall del clasificador optimizado: 0.409

Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.72      0.91      0.80      3762
           1       0.73      0.41      0.52      2238

    accuracy                           0.72      6000
   macro avg       0.73      0.66      0.66      6000
weighted avg       0.72      0.72      0.70      6000



La matriz de confucion nos muestra que:
* Tenemos 3424 TN - Verdaderos negativos: Son pagos clasificados  correctamente como 'No pago'
* Tenemos 338 FP - Falsos positivos: Pagos clasificaos incorrectamente como 'Pago'
* Tenemos 1323 FN - Falsos negativos: Pagos incorrectamente clasificados como 'No pago' 
* Tenemos 915 TP - Verdaderos positivos: Pago correctamente clasificados como 'Pago'

Podemos ver que tenemos buen desempeño para identificar los 'No pago' de manera correcta, pero bajo desempeño en indentificar los 'Pago' de manera correcta


* En cuanto a la precision tenemos un valor de 0.73 , lo que nos dice que el 73% de los pagos clasificados como 'Pago' son correctas.
* El recall tiene un valor de 0.41 para la clase 'Pago', lo que me indica que el modelo identifica correctamente el 41% de los valores 'Pago'
* Para el valor de F1 tenemos 0.52 para la clase 'Pago', indica un "equilibrio" entre la presicion y el recall
* El accuracy del modelo es de 0.72, estando en un valor no muy malo


Conclusion: Tenemos falencias a la hora de clasificar correctamente los valores 'Pago', se deberia ajustar para asi evitar falsos positivos

# Pregunta 5:
Construye ahora una red neuronal usando MLPClassifier de dos capas. Prueba diferente numero de neuronas por capa. ¿afecta al resultado?

In [ ]:
from sklearn.neural_network import MLPClassifier

# definimos diferentes numero de neuronas por capa
hidden_layer_sizes = [(10, 10), (20, 10), (50, 25), (120,60)]

for layers in hidden_layer_sizes:
    print(f"\nred neuronal con capas ocultas: {layers}")
    
    mlp = MLPClassifier(hidden_layer_sizes=layers, max_iter=1000, random_state=0)
    mlp.fit(x_train, y_train)
    
    # evaluamos el modelo
    y_pred = mlp.predict(x_test)
    accuracy_test = mlp.score(x_test, y_test)
    accuracy_train = mlp.score(x_train, y_train)
    print(f'Accuracy en train: {accuracy_train:.3f}')
    print(f"Accuracy en test: {accuracy_test:.3f}")

    # mostramos la matriz de confusión y el reporte de clasificación
    print("Matriz de Confusión:")
    print(confusion_matrix(y_test, y_pred))
    print("\nReporte de Clasificación:")
    print(classification_report(y_test, y_pred))


red neuronal con capas ocultas: (10, 10)
Accuracy en train: 0.888
Accuracy en test: 0.884
Matriz de Confusión:
[[3424  338]
 [ 359 1879]]

Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.91      0.91      0.91      3762
           1       0.85      0.84      0.84      2238

    accuracy                           0.88      6000
   macro avg       0.88      0.87      0.88      6000
weighted avg       0.88      0.88      0.88      6000


red neuronal con capas ocultas: (20, 10)
Accuracy en train: 0.894
Accuracy en test: 0.873
Matriz de Confusión:
[[3362  400]
 [ 365 1873]]

Reporte de Clasificación:
              precision    recall  f1-score   support

           0       0.90      0.89      0.90      3762
           1       0.82      0.84      0.83      2238

    accuracy                           0.87      6000
   macro avg       0.86      0.87      0.86      6000
weighted avg       0.87      0.87      0.87      6000


red neuronal 

Al observar los resultados, podemos identificar que a medida que aumentamos el numero de neuronas nuestro modelos va entrando en sobren ajuste, lo que nos indica que la red menos compleja (10,10) dos capas, de 10 neuronas cada una es la mas adecuada, el que mas generaliza: 
* Opteniendo los mejores resultado: 
    * TN = 3362
    * FP = 400
    * FN = 365
    * TP = 1873
    * Con un accuracy de 0.88 tanto para train como test

# Pregunta 6:
GradientBoostingClasifier es uno de los metodos de scikitlearn que mejor resultados suelen producir. Implementa un clasificador utilizando GradientBoostingClasifier considerando 0.05, 0.1, 0.15, 0.2, 0.25, 0.3 como learning rate.¿que learning rate produce un resultado optimo usando recall?

### 1. Implementando GridSearchCV

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import recall_score
from sklearn.model_selection import GridSearchCV

# valores para el learning_rate
param_grid = {'learning_rate': [0.05, 0.1, 0.15, 0.2, 0.25, 0.3]}

gbc = GradientBoostingClassifier(random_state=0) # con max_depth por defecto y n_estimators tambien debido a que posteriormente
# se optimizara el learning_rate con el metodo de validacion cruzada GridSearchCV

# GridSearchCV para optimizar el recall con 10 folds haciendo validación cruzada
grid_search_gbc = GridSearchCV(estimator=gbc,
                               param_grid=param_grid,
                               scoring='recall',
                               cv=10)

grid_search_gbc.fit(x_train, y_train)

# Mostramos el mejor learning_rate basado en recall
print(f"Mejor learning_rate: {grid_search_gbc.best_params_['learning_rate']}")
print(f"Mejor recall en validación: {grid_search_gbc.best_score_:.3f}")

# Evaluamos el recall en el conjunto de test y train
y_pred = grid_search_gbc.best_estimator_.predict(x_test)
y_pred_train = grid_search_gbc.best_estimator_.predict(x_train)
print(f'Recall en train es: {recall_score(y_pred_train, y_train):.3f}')
print(f'Recall en test: {recall_score(y_test, y_pred):.3f}')

Mejor learning_rate: 0.2
Mejor recall en validación: 0.799
Recall en train es: 0.859
Recall en test: 0.807


In [113]:
# para responder la ultima pregunta, imprimimos la matriz de confusión
conf_matrix_gbc = confusion_matrix(y_test, y_pred)
print("Matriz de Confusión del Gradient Boosting Classifier:")
print(conf_matrix_gbc)

Matriz de Confusión del Gradient Boosting Classifier:
[[3409  353]
 [ 433 1805]]


### 2. Implementado un bluque sin GridSearchCV
Esto para ver cada recall de learning rate

In [114]:
# Por otro lado me gustaria hacer un bucle para ver el recall de cada learning_rate
results = {}

for lrate in param_grid['learning_rate']:
    gbc = GradientBoostingClassifier(learning_rate=lrate, random_state=0) # con dejar max_depth por defecto y n_estimators
    gbc.fit(x_train, y_train)
    y_pred = gbc.predict(x_test)
    rec = recall_score(y_test, y_pred)
    recall_train = recall_score(y_train, gbc.predict(x_train))
    results[lrate] = rec
    print(f"Con Learning_rate ={lrate:.2f} El Recall en test: {rec:.3f} y en train: {recall_train:.3f}")

best_lrate = max(results, key=results.get)
print(f"\nEl Mejor learning_rate (por recall en test): {best_lrate} con recall={results[best_lrate]:.3f}")


Con Learning_rate =0.05 El Recall en test: 0.790 y en train: 0.786
Con Learning_rate =0.10 El Recall en test: 0.806 y en train: 0.807
Con Learning_rate =0.15 El Recall en test: 0.805 y en train: 0.808
Con Learning_rate =0.20 El Recall en test: 0.807 y en train: 0.817
Con Learning_rate =0.25 El Recall en test: 0.806 y en train: 0.815
Con Learning_rate =0.30 El Recall en test: 0.806 y en train: 0.825

El Mejor learning_rate (por recall en test): 0.2 con recall=0.807


El mejor learning rate es 0.2, tiene un score de 0.807 para test y en train un 0.817, es el learning que genera el modelo que mejor generaliza

# Pregunta 7:
Usando la matriz de confusion de los tres casos (GradientBoosting, MLP y K-NN) y derivando las metricas que te parezcan oportunas ¿Que clasificador de los 3 seleccionarias y porque?

* Seleccionaria el MLPClassifier de dos capas con la combinacion de 10 y 10 neuronas, ya que tiene un accuracy de 0.85 en test y 0.87 en train, lo que indica que no hay sobreajuste. Es el modelo que mas generaliza y parejo "equilibrado"
* Ademas MLPClassifier con 10 neuronas por capa muestra una matriz de confusion que indica correctamenta la mayoria de los casos positivos TP (1879) y TN (3424)
* Cabe mensionar que el modelo GradientBoosting que fue optimizado con GridSearchCV que entrego el learning rate 0.2 no esta nada mal, pero este tiene un pequeño descenso en el recall de test fente a MLPClassifier, aun asi se podria decir que es el segundo merjor

* De igual manera el modelo GradientBoosting se podria mejorar ajustando la profundidad (complejidad) de los arboles, de igual modo el numero de arboles y tomar nuevamente las metricas clave como recall, el accuracy y F1